### PR TITLE
Dev Tools: Add Language Picker for dev and wpcalypso environments

### LINKED ISSUE #69024: Dev Tools: Switch account language helper
For UI testing, it would be helpful to have an account language switcher in Dev Tools:

image

**Done is:**
* [x] It's possible to switch the language for an account from Dev Tools.

### PR DESCRIPTION
**Pull request by danielbachhuber**
Fixes #69024

## Proposed Changes

To make it easier for developers to vet interfaces across a variety of languages, I'd like to add a Language Picker to Dev Tools:

image

When you click on it, it loads a modal for you to select a language:

image

After you select a language, the page refreshes to load the new strings.

## Testing Instructions

1. Open 'Account Settings' in Dev Tools and choose a different language from the modal.
2. Verify your account switches to the specified language.
3. Verify nothing else breaks.

### CODE REVIEW

File: client/lib/account-settings-helper/index.jsx [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+import getUserSettings from 'calypso/state/selectors/get-user-settings';
+import { setUserSetting } from 'calypso/state/user-settings/actions';
+import { saveUnsavedUserSettings } from 'calypso/state/user-settings/thunks';
+
+import './style.scss';
+
+export function AccountSettingsHelper() {
+	const dispatch = useDispatch();
+	const userSettings = useSelector( getUserSettings ) ?? {};
+	const updateLanguage = ( event ) => {
```
Comment (danielbachhuber) [inline client/lib/account-settings-helper/index.jsx:15]: `/me/account` has a bunch of additional logic around `updateLanguage`:

https://github.com/Automattic/wp-calypso/blob/6e0933e052ac8d9aa6ec881208d31a3446a5df5a/client/me/account/main.jsx#L163-L203

What else of this is necessary to recreate for this particular context?

**Comment by zaguiini** [comment]
Is it possible to:

1. avoid a full page refresh?
2. change the language only for the current page?

Usually we only want to test the change and that's temporary, so it doesn't make much sense to persist this across refreshes 🤔

**Comment by vykes-mac** [comment]
I appreciate this feature @danielbachhuber , helps in spotting UI incosistencies for different languages without leaving the page. I would be nice to have what @zaguiini mentioned but seems would involve good amount of effort.

File: client/document/index.jsx [hunk 2]
```diff
@@ -154,6 +156,7 @@ class Document extends Component {
 					{ badge && (
 						<EnvironmentBadge badge={ badge } feedbackURL={ feedbackURL }>
 							{ reactQueryDevtoolsHelper && <ReactQueryDevtoolsHelper /> }
+							{ accountSettingsHelper && <AccountSettingsHelper /> }
```
Comment (sejas) [inline client/document/index.jsx:159]: 👌

**Review by sejas — COMMENTED** [review]
I like the idea of adding new features to the dev menu. It's a very powerful idea. Thank you!!
It works as described.
I agree with Luis. I would prefer if the language change is temporal.

**Comment by lynnbgriffith** [comment]
> 1. avoid a full page refresh?

In general, I think this should be possible, but certain calls like `i18n.translate` will not update/re-render like you'd expect. I believe @jsnajdr has worked in that space previously, so he might have some thoughts.

**Comment by danielbachhuber** [comment]
> Is it possible to:
> 
>     1. avoid a full page refresh?
> 
>     2. change the language only for the current page?
> 
> Usually we only want to test the change and that's temporary, so it doesn't make much sense to persist this across refreshes 🤔

I'm not sure I'm a huge fan of this idea:
* It would be a departure from the other Dev Tool menu items (which persist their settings) and Calypso generally.
* If I switched languages to fix something for a specific language, it would be annoying to have to switch back to the language after each refresh.

Also, there's a lot of complexity around localization so I think it's worth following the happy path for this particular limited use case.

**Comment by zaguiini** [comment]
> > Is it possible to:
> > ```
> > 1. avoid a full page refresh?
> > 
> > 2. change the language only for the current page?
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Usually we only want to test the change and that's temporary, so it doesn't make much sense to persist this across refreshes 🤔
> 
> I'm not sure I'm a huge fan of this idea:
> 
> * It would be a departure from the other Dev Tool menu items (which persist their settings) and Calypso generally.
> * If I switched languages to fix something for a specific language, it would be annoying to have to switch back to the language after each refresh.
> 
> Also, there's a lot of complexity around localization so I think it's worth following the happy path for this particular limited use case.

Fair enough 👍

**Comment by mpkelly** [comment]
Tested successfully on Chrome and Firefox. Nice addition!

You gave me an idea: listen for Console errors and show a counter in the Dev Tools, which could change to red if errors are logged. 

Does anyone think this would be useful and worthwhile?

**Review by wojtekn — APPROVED** [review]
@danielbachhuber works well for me, nice addition.

File: client/lib/account-settings-helper/index.jsx (hunk 1 above)
Comment (danielbachhuber) [reply client/lib/account-settings-helper/index.jsx:15]:
```
if ( typeof empathyMode !== 'undefined' ) { 
 	this.updateUserSetting( 'i18n_empathy_mode', empathyMode ); 
}
```

I don't think `empathyMode` is totally relevant for our use case (#37106) but I've added support anyway: fd3ae3e725

```
if ( typeof useFallbackForIncompleteLanguages !== 'undefined' ) { 
	this.updateUserSetting( 
		'use_fallback_for_incomplete_languages', 
		useFallbackForIncompleteLanguages 
	); 
} 
```
`useFallbackForIncompleteLanguages` seems to be necessary and useful (#43571). Added support (and fixed bug in other changeset): 08cc3e4a4e

The other stuff seems related to the `/me/account` UI.

**Comment by danielbachhuber** [comment]
> Tested successfully on Chrome and Firefox. Nice addition!

@mpkelly Great, thanks!

> You gave me an idea: listen for Console errors and show a counter in the Dev Tools, which could change to red if errors are logged.
> 
> Does anyone think this would be useful and worthwhile?

Personally, I tend to have my Console open all of the time anyway (because my JavaScript often causes errors 🙃)

You could take a pass at incorporating a `Shema boja nadzorne ploče` into the tool:

image

I might use a pure `<select>` though, as those visual previews might be too much to fit in the space allotted.

File: client/components/environment-badge/index.jsx [hunk 3]
```diff
@@ -10,6 +10,12 @@ export function ReactQueryDevtoolsHelper() {
 	/* eslint-enable wpcalypso/jsx-classname-namespace */
 }
 
+export function AccountSettingsHelper() {
+	/* eslint-disable wpcalypso/jsx-classname-namespace */
+	return <div className="environment is-account-settings" />;
+	/* eslint-enable wpcalypso/jsx-classname-namespace */
+}
```
Comment (jsnajdr) [inline client/components/environment-badge/index.jsx:17]: What about calling the helper just `<LangHelper>`? It changes only language and is unlikely to do anything else.

The `eslint-disable` comments are no longer needed these days. The `wpcalypso/jsx-classname-namespace` rule has changed a few months ago and now accepts more classnames.

File: client/lib/account-settings-helper/index.jsx (hunk 1 above)
Comment (jsnajdr) [reply client/lib/account-settings-helper/index.jsx:15]: For a dev helper for quickly switching languages, you probably want to copy the `QuickLanguageSwitcher` component as much as possible. It only calls the `setLocale` Redux action, setting the language for this particular session, overriding the user setting. And it doesn't persist the language, on reload you'll be back in the default language.

**Review by jsnajdr — COMMENTED** [review]
After adding the new feature, the `QuickLanguageSwitcher` UI seems to be obsolete: it performs the same function, just in a different place in the UI (masterbar), and is behind a feature flag.

**Comment by jsnajdr** [comment]
> In general, I think this should be possible, but certain calls like `i18n.translate` will not update/re-render like you'd expect. I believe @jsnajdr has worked in that space previously, so he might have some thoughts.

Yes, non-reactive translations done with a `i18n.translate` call won't be updated on `setLocale`. But for a dev helper, that's acceptable, and also adds an incentive to migrate the offending code away 🙂 

Anyway, exactly the same functionality is already implemented in `QuickLanguageSwitcher`, this PR just re-implements it in a different UI place. I'm wondering if @danielbachhuber was aware of it when working on this?

File: client/components/environment-badge/index.jsx (hunk 3 above)
Comment (danielbachhuber) [reply client/components/environment-badge/index.jsx:17]: > What about calling the helper just `<LangHelper>`? It changes only language and is unlikely to do anything else.

@jsnajdr To create space for it to also switch the `Dashboard Color Scheme`.

> The `eslint-disable` comments are no longer needed these days. The `wpcalypso/jsx-classname-namespace` rule has changed a few months ago and now accepts more classnames.

Ok, removed them in https://github.com/Automattic/wp-calypso/pull/69073

**Comment by danielbachhuber** [comment]
@jsnajdr Where does this `QuickLanguageSwitcher` appear? If an entire team isn't aware of it, maybe it's not accessible enough?

> For a dev helper for quickly switching languages, you probably want to copy the `QuickLanguageSwitcher` component as much as possible. It only calls the `setLocale` Redux action, setting the language for this particular session, overriding the user setting. And it doesn't persist the language, on reload you'll be back in the default language.

Per my comment above (https://github.com/Automattic/wp-calypso/pull/69026#issuecomment-1278256433), my preference is persistent:

> I'm not sure I'm a huge fan of this idea:
> * It would be a departure from the other Dev Tool menu items (which persist their settings) and Calypso generally.
> * If I switched languages to fix something for a specific language, it would be annoying to have to switch back to the language after each refresh.

File: client/components/environment-badge/index.jsx (hunk 3 above)
Comment (jsnajdr) [reply client/components/environment-badge/index.jsx:17]: > To create space for it to also switch the Dashboard Color Scheme.

Oh yes, that's a neat idea 🙂

**Comment by jsnajdr** [comment]
> Where does this `QuickLanguageSwitcher` appear?

It appears in the masterbar:

Screenshot 2022-10-14 at 14 43 08

either if you are inside of a support session, or when you have the `quick-language-switcher` feature flag enabled, by adding a `?flags=...` param to the Calypso URL.

> If an entire team isn't aware of it, maybe it's not accessible enough?

It's primarily a Happiness tool (I forgot about the support session condition when writing the previous comment). As a dev tool, yes, it's only an afterthough and it's hard to find and activate.

Both ways to switch languages continue to be useful and valid. Nothing gets obsolete, I was wrong.

> my preference is persistent:

OK, both ways make sense in a dev tool. On the other hand, a happiness tool must not change the user preference.

**Comment by danielbachhuber** [comment]
> It's primarily a Happiness tool (I forgot about the support session condition when writing the previous comment). As a dev tool, yes, it's only an afterthough and it's hard to find and activate.

@jsnajdr Oh, I actually think I was aware of it in this context. However, given it was support session-specific (or at least I thought so), I thought it would be helpful to have something comparable for development.

> Both ways to switch languages continue to be useful and valid. Nothing gets obsolete, I was wrong.

Cool, makes sense. Glad we were able to work this through 😊

**Comment by mpkelly** [comment]
> You could take a pass at incorporating a `Shema boja nadzorne ploče` into the tool:

Challenge accepted!

**Comment by danielbachhuber** [comment]
> Challenge accepted!

@mpkelly Awesome! I created an issue for you #69084

**Pull request merged**